granulate_utils/linux/cgroups/cgroup.py

```diff
@@ -73,28 +95,231 @@ def find_v2_hierarchy(resolve_host_root_links: bool = True) -> Optional[str]:
-    the v2 mountpoint is returned (as all controllers share the same hierarchy in v2)
+    a CgroupCoreV2 is returned (as all controllers share the same hierarchy in v2)
     """
-    v1_paths = find_v1_hierarchies(resolve_host_root_links)
-    v2_path = find_v2_hierarchy(resolve_host_root_links)
-    if v1_paths and controller in v1_paths:
+    v1_paths = _find_v1_hierarchies()
+    v2_path = _find_v2_hierarchy()
+    if controller in v1_paths:
         # Either v1 or hybrid with the requested controller bound to a v1 hierarchy
-        return v1_paths[controller]
+        return CgroupCoreV1(Path(v1_paths[controller]), Path(v1_paths[controller]))
     if v2_path:
-        # v2 (unified) - all controllers are in a single unified hierarchy
-        return v2_path
+        # v2 (unified) - check given controller is supported in cgroup v2
+        cgroup_v2 = CgroupCoreV2(Path(v2_path), Path(v2_path))
+        if cgroup_v2.is_controller_supported(controller):
+            return cgroup_v2
     # No cgroup mount of the requested controller
     return None
 
 
-def is_known_controller(controller: str) -> bool:
-    return controller in SUBSYSTEMS
+def _get_cgroup_mount_checked(controller: ControllerType) -> CgroupCore:
+    cgroup_mount = _get_cgroup_mount(controller)
+    if cgroup_mount is None:
+        raise CgroupControllerNotMounted(controller_name=controller)
+    return cgroup_mount
+
+
+def _get_cgroup_from_path(controller: ControllerType, cgroup_path_or_full_path: Path) -> CgroupCore:
+    cgroup_mount = _get_cgroup_mount_checked(controller)
+
+    try:
+        cgroup_path_or_full_path.relative_to(cgroup_mount.cgroup_abs_path)
+        # it's a full path
+        cgroup_abs_path = cgroup_path_or_full_path
+    except ValueError:
+        cgroup_path_or_full_path = Path(cgroup_path_or_full_path.as_posix().lstrip("/"))
+        # it's a path relative to controller mount point
+        cgroup_abs_path = cgroup_mount.cgroup_abs_path / cgroup_path_or_full_path
+
+    return cgroup_mount.build_object(cgroup_abs_path, cgroup_mount.cgroup_mount_path)
+
+
+def _get_cgroup_for_process(controller: ControllerType, process: Optional[psutil.Process] = None) -> CgroupCore:
+    """
+    Get a CgrouopCore object for a given process. If process is None return for current process.
+    """
+    cgroup_mount = _get_cgroup_mount_checked(controller)
+
+    for process_cgroup in get_process_cgroups(process):
+        if (cgroup_mount.is_v1 and controller in process_cgroup.controllers) or (
+            cgroup_mount.is_v2 and process_cgroup.hier_id == "0"
+        ):
+            # Validate hier == 0 only when cgroup is v2
+            assert cgroup_mount.is_v1 ^ (process_cgroup.hier_id == "0")
+            return cgroup_mount.build_object(
+                cgroup_mount.cgroup_abs_path / process_cgroup.relative_path.lstrip("/"), cgroup_mount.cgroup_mount_path
+            )
```

Comment:
Why do we need to check `hier_id == 0`?

Reply:
I removed the `assert`, but we need that check since in cgroupv2 the only indication we have that the cgroup line in `/proc/{pid}/cgroup` is the unified cgroup is that the `hier_id == 0`